File: tunnelling_probability.py
import math

class QuantumTunnelling:
    """ Planck's constant """
    K_PLANK = 6.626e-34
    
    """ MASS OF ELECTRON (kg) """
    K_ELECTRON_MASS = 9.1e-31
    
    """ Electron volt """
    K_EV = 1.6e-19
    
    @staticmethod
    def EngelProbability(size, E, V):
        """
        Engel's Quantum Tunnelling Probability
        :param size: size of barrier in meters.
        :param E: kinetic energy in electron volt (eV)
        :param V: potential energy in eV.
        :return: Quantum Tunnelling Probability
        """ 
        if E > V:
            raise Exception("Potential energy (V) must be > Kinetic Energy (E)")
        delta = V - E
        p1 = ((-4 * size * math.pi) / QuantumTunnelling.K_PLANK)
        p2 = math.sqrt(2 * QuantumTunnelling.K_ELECTRON_MASS * delta * QuantumTunnelling.K_EV)
        return math.exp(p1 * p2)
    
def main():
    """ A simple test for current semiconductor processes """
    try:
        # Barrier sizes for current semiconductor processes (m)
        SIZES = [130e-9, 32e-9, 14e-9, 7e-9, 5e-9, 500e-12]
        # Dates for display purposes
        DATES = ["2001", "2010", "2014", "2019", "2021", "Beyond"]
        E = 4.5 # Kinetic energy of electron (eV)
        V = 5.0 # Potential Energy (eV)
        
        # Display them
        for i in range(len(DATES)): 
            p = QuantumTunnelling.EngelProbability(SIZES[i], E, V)
            print("{}\t{:.2e}\t{:.3e}".format(DATES[i], SIZES[i], p))
    
    except Exception as e:
        print(e)

if __name__ == "__main__":
    main()

File: tunnelling_probability.java
public class Quantum Tunnelling {
    /** Planck's constant */
    static final double K_PLANK = 6.626e-34;

    /** MASS OF ELECTRON (kg) */
    static final double K_ELECTRON_MASS = 9.1e-31;

    /** Electron volt */
    static final double K_EV = 1.6e-19;

    /**
    * Engel's Quantum Tunnelling Probability
    * @param size
    * size of barrier in meters.
    * @param E
    * kinetic energy in electron volt (eV)
    * @param V
    * potential energy in eV.
    * @return Quantum Tunnelling Probability
    */ 

    static double EngelProbability(double size, double E, double V) {
        if (E > V) {
            throw new IllegalArgumentException
            ("Potential energy (V) must be > Kinetic Energy (E)");
        }
        double delta = V - E;
        double p1 = ((-4 * size * Math.PI) / K_PLANK);
        double p2 = Math.sqrt(2 * K_ELECTRON_MASS * delta * K_EV);
        return Math.exp(p1 * p2);
    }

    /** A simple test for current semiconductor processes */
    public static void main(String[] args){
        try {
            //Barrier sizes for current semiconductor processes (m)
            final double[] SIZES = { 130-9, 32e-9, 14e-9, 7e-9, 5e-9, 500e-12 };
            //Dates for display purposes
            final String[] DATES = { "2001", "2010", "2014", "2019", "2021", "Beyond"};
            final double E = 4.5; //Kinetic energy of electron (eV)
            final double V = 5.0; //Potential Energy (eV)

            //dISPLAY THEM::
            for (int in = 0; i < DATES.length; i++) { 
                double p = Engel Probability(SIZES[i], E, V);
                System.out.println(String.format("%s\t%2.2e\t%2.3e",
                DATES[i], SIZES[i], p));
            }

        } catch (Expection e) {
            e.printStackTrace();
        }
    }
}
